smm/lotus_adapter.go

```diff
@@ -0,0 +1,266 @@
+        From:     adapter.address,
+        Method:   method,
+        Params:   payload,
+        // TODO: Add ability to control these 'costs'
+        Value:    types.NewInt(1000), // currently hard-coded late fee in actor, returned if not late
+        GasLimit: types.NewInt(1000000 /* i dont know help */),
+        GasPrice: types.NewInt(1),
+    }
+
+    var msgErr error
+    signedMsg, msgErr := adapter.fullNode.MpoolPushMessage(ctx, &msg)
+
+    if msgErr != nil {
+        err = msgErr
+        return
+    }
+    cid = signedMsg.Cid()
+    return
+}
+
+func (adapter lotusAdapter) MostRecentState(ctx context.Context) (stateKey *StateKey, epoch Epoch, err error) {
+    var ts *types.TipSet
+    ts, err = adapter.fullNode.ChainHead(ctx)
+    if err != nil {
+        return
+    }
+    epoch = Epoch(ts.Height())
+    stateKey, err = tipset2statekey(ts)
+    return
+}
+
+func (adapter lotusAdapter) GetMinerState(ctx context.Context, stateKey *StateKey) (state MinerChainState, err error) {
+    var ts *types.TipSet
+    ts, err = statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return
+    }
+    sectors, apiErr := adapter.fullNode.StateMinerSectors(ctx, adapter.address, ts)
+    if apiErr != nil {
+        err = apiErr
+        return
+    }
+    provingSectors, apiErr := adapter.fullNode.StateMinerProvingSet(ctx, adapter.address, ts)
+    if apiErr != nil {
+        err = apiErr
+        return
+    }
+    state.Sectors = make(map[uint64]api.ChainSectorInfo)
+    for _, sectorInfo := range sectors {
+        state.Sectors[sectorInfo.SectorID] = *sectorInfo
+    }
+    for _, sectorInfo := range provingSectors {
+        state.ProvingSet[sectorInfo.SectorID] = struct{}{}
+    }
+    state.Address = Address(adapter.address.String())
+    // fill in this information
+    state.PreCommittedSectors = make(map[uint64]api.ChainSectorInfo)
+    state.StagedCommittedSectors = make(map[uint64]api.ChainSectorInfo)
+
+    return
+}
+
+func (adapter lotusAdapter) GetRandomness(ctx context.Context, stateKey *StateKey, e Epoch, offset uint) (buffer []byte, err error) {
+    var ts *types.TipSet
+    ts, err = statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return
+    }
+    buffer, err = adapter.fullNode.ChainGetRandomness(ctx, ts, ts.Blocks()[0].Tickets, int(offset))
+    return
+}
+
+func (adapter lotusAdapter) GetProvingPeriod(ctx context.Context, stateKey *StateKey) (pp ProvingPeriod, err error) {
+    var ts *types.TipSet
+    ts, err = statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return
+    }
+    end, apiErr := adapter.fullNode.StateMinerProvingPeriodEnd(ctx, adapter.address, ts)
+    if apiErr != nil {
+        err = apiErr
+        return
+    }
+    pp.End = Epoch(end)
+    pp.Start = Epoch(end - build.ProvingPeriodDuration)
+    // Not sure what goes in these fields
+    _ = pp.ChallengeSeed
+    _ = pp.ChallengeStart
+    _ = pp.ChallengeEnd
```

Comment:
These are for surprise post, which Lotus doesn't implement (yet). So change the types to reflect the windowed rational post that's actually implemented (i.e. remove these three)